Bottom: adc261b4f2545a74209a3f6d8ffebb209f129502
Top:    1435046ba8704c82bcb6e5a4a4d89d1f189389c3
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2012-12-23 02:07:17 +0000

Refresh of major-re-write-to-drawing-rout

---

diff --git a/src/hid/gtk/gtkhid-gl.c b/src/hid/gtk/gtkhid-gl.c
index 151a815..b2b48c8 100644
--- a/src/hid/gtk/gtkhid-gl.c
+++ b/src/hid/gtk/gtkhid-gl.c
@@ -927,10 +927,6 @@ ghid_start_drawing (GHidPort *port, GtkWidget *widget)
 
   port->render_priv->in_context = true;
 
-  Output.fgGC = gui->graphics->make_gc ();
-  Output.bgGC = gui->graphics->make_gc ();
-  Output.pmGC = gui->graphics->make_gc ();
-
   return TRUE;
 }
 
@@ -948,14 +944,6 @@ ghid_end_drawing (GHidPort *port, GtkWidget *widget)
 
   /* end drawing to current GL-context */
   gdk_gl_drawable_gl_end (pGlDrawable);
-
-  gui->graphics->destroy_gc (Output.fgGC);
-  gui->graphics->destroy_gc (Output.bgGC);
-  gui->graphics->destroy_gc (Output.pmGC);
-
-  Output.fgGC = NULL;
-  Output.bgGC = NULL;
-  Output.pmGC = NULL;
 }
 
 void
@@ -1716,6 +1704,10 @@ ghid_drawing_area_expose_cb (GtkWidget *widget,
   ghid_start_drawing (port, widget);
   hidgl_start_render ();
 
+  Output.fgGC = gui->graphics->make_gc ();
+  Output.bgGC = gui->graphics->make_gc ();
+  Output.pmGC = gui->graphics->make_gc ();
+
   /* If we don't have any stencil bits available,
      we can't use the hidgl polygon drawing routine */
   /* TODO: We could use the GLU tessellator though */
@@ -1896,6 +1888,13 @@ ghid_drawing_area_expose_cb (GtkWidget *widget,
   hidgl_finish_render ();
   ghid_end_drawing (port, widget);
 
+  gui->graphics->destroy_gc (Output.fgGC);
+  gui->graphics->destroy_gc (Output.bgGC);
+  gui->graphics->destroy_gc (Output.pmGC);
+
+  Output.fgGC = NULL;
+  Output.bgGC = NULL;
+  Output.pmGC = NULL;
   g_timer_start (priv->time_since_expose);
 
   return FALSE;
